### PR TITLE
use pytorch data loader to train

### PR DESCRIPTION
**Pull request by dpressel**
Use PyTorch's built-in `DataLoader` to traverse the MEAD-generated map-style dataset.
This is a pre-cursor for doing larger-scale processing in the future with access to multiple workers.

Analysis of the timing logs shows that the impact is negligable, and the model performance is not impacted.

One change that is required is to not assume that `make_input` should be accessing numpy arrays,
I added a `numpy_to_tensor` boolean to the argument list and to the sub-methods it calls for each task, and default this to false, so that for all training we do not have it on.

I also made the predict methods support `kwargs` and default to `numpy_to_tensor` on.

### CODE REVIEW

File: baseline/pytorch/classify/model.py [hunk 1]
```diff
@@ -54,46 +54,76 @@ def cuda(self, device=None):
...
+        perm_idx = None
 
         # Allow us to track a length, which is needed for BLSTMs
         if self.lengths_key is not None:
-            example_dict['lengths'] = torch.from_numpy(batch_dict[self.lengths_key])
+            lengths = batch_dict[self.lengths_key]
+
+            if numpy_to_tensor:
+                lengths = torch.from_numpy(lengths)
+                lengths, perm_idx = lengths.sort(0, descending=True)
```
Comment (blester125) [inline baseline/pytorch/classify/model.py]: This part should be outside of this if right? We can still have the situation where we want to sort the lengths even if using a dataset right?

File: baseline/pytorch/lm/train.py [hunk 2]
```diff
@@ -46,15 +47,16 @@ def _get_pytorch_model(self):
         return self.model.module if self.gpus > 1 else self.model
 
     @staticmethod
-    def _get_dims(batch_dict):
+    def _get_dims(loader):
```
Comment (blester125) [inline baseline/pytorch/lm/train.py]: Not related to this PR because the problem is already there but some of the tf trainers call this `_get_batchsz` we should be consistent

Comment (blester125) [reply baseline/pytorch/lm/train.py]: nvm I see now that that it can be used for dims other than batch size like when getting the number of tokens

File: baseline/pytorch/lm/train.py [hunk 3]
```diff
@@ -46,15 +47,16 @@ def _get_pytorch_model(self):
...
+        batch_dict = loader.dataset[0]
         return batch_dict['y'].shape
 
     @staticmethod
     def _num_toks(batch_dict):
-        return np.prod(LanguageModelTrainerPyTorch._get_dims(batch_dict))
+        return np.prod(batch_dict['y'].shape)
```
Comment (blester125) [inline baseline/pytorch/lm/train.py]: I see that we don't use the `_get_dims` anymore because this `_num_toks` need to be dynamic (needs to report for each batch because it can be different) but that raises the question of if we can remove the `_get_dims` function.

Also although we always uses dense LM batches would it make sense to calculate the number of tokens based on the lengths in case we ever need a LM that has padding (like each example if a single sentence)

File: baseline/pytorch/classify/model.py (hunk 1 above)
Comment (dpressel) [reply baseline/pytorch/classify/model.py]: yep

File: baseline/pytorch/seq2seq/train.py [hunk 4]
```diff
@@ -43,7 +44,7 @@ def __init__(self, model, **kwargs):
 
     @staticmethod
     def _num_toks(tgt_lens):
-        return np.sum(tgt_lens)
+        return float(torch.sum(tgt_lens))
```
Comment (blester125) [inline baseline/pytorch/seq2seq/train.py]: I thought the suggested way to get the value from a pytorch tensor into a python number was to use `.item`?

Also the number of tokens seems like it should be an int not a float?

File: baseline/pytorch/classify/model.py (hunk 1 above)
Comment (dpressel) [reply baseline/pytorch/classify/model.py]: fixed

File: baseline/pytorch/seq2seq/train.py (hunk 4 above)
Comment (dpressel) [reply baseline/pytorch/seq2seq/train.py]: We could.  AFAICT its only ever used to divide out tokens or multiply by another float

File: baseline/pytorch/lm/train.py (hunk 3 above)
Comment (dpressel) [reply baseline/pytorch/lm/train.py]: We currently use `_get_dims` in cases where we are doing truncated backprop, and its a bit unfortunate that we get the batch size this way as it makes the code clunky and not very robust -- it ends up assuming that we have a map-style dataset which we dont want to assume probably moving forward.

File: baseline/pytorch/seq2seq/train.py (hunk 4 above)
Comment (dpressel) [reply baseline/pytorch/seq2seq/train.py]: switched to item
